v2/writer.go

```diff
@@ -210,3 +214,97 @@ func AttachIndex(path string, idx index.Index, offset uint64) error {
 	indexWriter := internalio.NewOffsetWriter(out, int64(offset))
 	return index.WriteTo(idx, indexWriter)
 }
+
+// ReplaceRootsInFile replaces the root CIDs in CAR file at given path with the given roots.
+// This function accepts both CARv1 and CARv2 files.
+//
+// Note that the roots are only replaced if their total serialized size exactly matches the total
+// serialized size of existing roots in CAR file.
+// This function does not shift sections in CAR file back or forth in order fit the given roots.
+func ReplaceRootsInFile(path string, roots []cid.Cid) (err error) {
+	f, err := os.OpenFile(path, os.O_RDWR, 0o666)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		// Close file and override return error type if it is nil.
+		if cerr := f.Close(); err == nil {
+			err = cerr
+		}
+	}()
+
+	// Read header or pragma; note that both are a valid CARv1 header.
+	header, err := carv1.ReadHeader(f)
+	if err != nil {
+		return err
+	}
+
+	var currentSize int64
+	var newHeaderOffset int64
+	switch header.Version {
+	case 1:
+		// When the given file is a CARv1 :
+		// 1. The offset at which the new header should be written is zero (newHeaderOffset = 0)
+		// 2. The current header size is equal to the number of bytes read, and
+		//
+		// Note that we explicitly avoid using carv1.HeaderSize to deremine the current header size.
+		// This is based on the fact that carv1.ReadHeader does not read any extra bytes.
+		// Therefore, we can avoid extra allocations of carv1.HeaderSize to determine size by simply
+		// counting the bytes read so far.
+		currentSize, err = f.Seek(0, io.SeekCurrent)
+		if err != nil {
+			return err
+		}
+	case 2:
+		// When the given file is a CARv2 :
+		// 1. The offset at which the new header should be written is carv2.Header.DataOffset
+		// 2. The inner CARv1 header size is equal to the value of varint read at carv2.Header.DataOffset,
+		//    plus the varint size itself.
+		//
+		// By reading the varint we avoid parsing the inner CARv1 header to determine its size.
+		// Its content is not important since it'll be replaced.
+		var v2h Header
+		if _, err = v2h.ReadFrom(f); err != nil {
+			return err
+		}
+		newHeaderOffset = int64(v2h.DataOffset)
+		if _, err = f.Seek(newHeaderOffset, io.SeekStart); err != nil {
+			return err
+		}
+		var innerSize uint64
+		innerSize, err = varint.ReadUvarint(internalio.ToByteReader(f))
+		if err != nil {
+			return err
+		}
+		currentSize = int64(innerSize + uint64(varint.UvarintSize(innerSize)))
```

Comment:
I wonder if it would be simpler to just do the same logic as in the carv1? after all, we just need to do exactly the same stuff, but with the inner section if it's a carv2.

Reply:
another reason to always read the "true" inner carv1 header is to ensure that it's valid. right now, if a carv2 has an inner carv1 header with version=5, you'll happily replace it with version=1 without erroring. instead we should likely complain.